src/nbl/asset/utils/CSPIRVIntrospector.cpp

```diff
@@ -13,472 +13,528 @@ namespace nbl::asset
-{
-    std::pair<bool, IImageView<ICPUImage>::E_TYPE> fail = { false, IImageView<ICPUImage>::ET_COUNT };
-#if 0 // rewrite for sanity
-    core::vector<core::smart_refctd_ptr<const CIntrospectionData>> introspections(infos.size());
-    if (!introspectAllShaders(introspections.data(),infos))
-        return fail;
-
-    for (auto i=0; i<infos.size(); i++)
-    for (const auto& bnd : introspections[i]->descriptorSetBindings[_set])
-    switch (bnd.type)
-    {
-        case ESRT_COMBINED_IMAGE_SAMPLER:
-            return imageInfoFromResource(bnd.get<ESRT_COMBINED_IMAGE_SAMPLER>());
-            break;
-        case ESRT_STORAGE_IMAGE:
-            return imageInfoFromResource(bnd.get<ESRT_STORAGE_IMAGE>());
-            break;
-        default:
-            break;
-    }
-#endif
-    return fail;
-}
+                    const auto& specConstantFound = specConstants->find(arrInfo.specID);
+                    if (specConstantFound == specConstants->end())
+                        return 0u;
 
-core::smart_refctd_dynamic_array<SPushConstantRange> CSPIRVIntrospector::createPushConstantRangesFromIntrospection_impl(
-    core::smart_refctd_ptr<const CIntrospectionData>* const introspections,
-    const std::span<const ICPUShader::SSpecInfo> infos)
-{
-#if 0 // rewrite for sanity
-    core::vector<core::vector<SPushConstantRange>> ranges(infos.size());
-    {
-        auto introspection = introspections;
-        auto r = ranges.begin();
-        for (auto info : infos)
+                    return val * specConstantFound->second;
+                }
+
+                return val * arrInfo.value;
+            };
+
+        const auto& introBindingInfos = stageData->getDescriptorSetInfo(i);
+        for (const auto& stageIntroBindingInfo : introBindingInfos)
         {
-            auto& pc = (*introspection)->pushConstant;
-            if (pc.present)
+            CPipelineIntrospectionData::SDescriptorInfo descInfo;
+            descInfo.binding = stageIntroBindingInfo.binding;
+            descInfo.type = stageIntroBindingInfo.type;
+            descInfo.stageMask = stageData->m_shaderStage;
+            if (stageIntroBindingInfo.isArray())
             {
-                const auto shaderStage = info.shader->getStage();
-
-                r->reserve(100u);
-                auto& members = pc.info.members;
-                r->push_back({shaderStage, members.array[0].offset, members.array[0].size});
-                for (uint32_t i = 1u; i < members.count; ++i)
+                if (stageIntroBindingInfo.isRunTimeSized())
+                {
+                    descInfo.count = 0u;
+                    descInfo.stride = std::accumulate(stageIntroBindingInfo.count.begin(), stageIntroBindingInfo.count.end(), 1u, multiplyArrayDimensions);
+                    if (descInfo.stride == 0u)
+                        return false;
```

Comment:
maybe since the check is the same in both branches, you want to hoist it out and move it after the if-else branch